File: src/structural/decorator.java
public class DatabaseService {
    public String getMailFromUsername(String username) {
        return username + "@mail";
    }
}

public class Notifier {
    private final String username;
    protected final DatabaseService databaseService;

    public Notifier(String username) {
        this.username = username;
        databaseService = new DatabaseService();
    }

    public void send(String msg) {
        String mail = databaseService.getMailFromUsername(username);
        System.out.println("Sending " + msg + " by Mail to " + mail);
    }

    public String getUsername() {
        return username;
    }
}

public class DatabaseService {
    public String getPhoneNbrFromUsername(String username) {
        return username + "@Phone";
    }

    public String getFBNameFromUsername(String username) {
        return username + "@Facebook";
    }
}

public class WhatsAppNotifier extends Notifier {
    public WhatsAppNotifier(String username) {
        super(username);
    }

    @Override
    public void send(String msg) {
        String phoneNbr = databaseService.getPhoneNbrFromUsername(getUsername());
        System.out.println("Sending " + msg + " by WhatsApp on " + phoneNbr);
    }
}

public class FacebookNotifier extends Notifier {
    public FacebookNotifier(String username) {
        super(username);
    }

    @Override
    public void send(String msg) {
        String fbName = databaseService.getFBNameFromUsername(getUsername());
        System.out.println("Sending " + msg + " on Facebook to " + fbName);
    }
}

public static void main(String[] args) {
    WhatsAppNotifier notifier = new WhatsAppNotifier("Geekific");
    notifier.send("Like and Subscribe!");
}

File: src/structural/decorator.ts
class InitialDatabaseService {
  getMailFromUsername(username: string): string {
    return username + "@mail";
  }
}

class Notifier {
  protected databaseService = new DatabaseService();

  constructor(private username: string) {}

  send(msg: string): void {
    const mail: string = this.databaseService.getMailFromUsername(this.username);
    console.log("Sending " + msg + " by Mail to " + mail);
  }

  getUsername(): string {
    return this.username;
  }
}

class DatabaseService extends InitialDatabaseService {
  getPhoneNbrFromUsername(username: string): string {
    return username + "@Phone";
  }

  getFBNameFromUsername(username: string): string {
    return username + "@Facebook";
  }
}

class WhatsAppNotifier extends Notifier {
  constructor(username: string) {
    super(username);
  }

  override send(msg: string): void {
    const phoneNbr: string = this.databaseService.getPhoneNbrFromUsername(this.getUsername());
    console.log("Sending " + msg + " by WhatsApp on " + phoneNbr);
  }
}

class FacebookNotifier extends Notifier {
  constructor(username: string) {
    super(username);
  }

  override send(msg: string): void {
    const fbName = this.databaseService.getFBNameFromUsername(this.getUsername());
    console.log("Sending " + msg + " on Facebook to " + fbName);
  }
}

function sendWhatsAppNotification(): void {
  const notifier = new WhatsAppNotifier("Geekific");
  notifier.send("Like and Subscribe!");
}

function sendFBNotification(): void {
  const notifier = new FacebookNotifier("Geekific");
  notifier.send("Like and Subscribe!");
}

sendWhatsAppNotification();
sendFBNotification();
